### PR TITLE
Add Dockerfile for Graph

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

It's easier to only run the docker container than to build the Rust projects, especially if you are not used to Rust. This PR adds a Dockerfile and builds it in the CI.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1202771673621395) _(internal)_

## 🔍 What does this change?

- Add a Dockerfile
- Add commands to the Makefile
- Add a new CI step to build the docker container
  - We chose to not cache the layers of the docker build as the build is pretty fast. With caching we would require a caching infrastructure for Docker, and also upload/download will take a while. In addition, docker layers tend to be big. The final Docker container however is designed to be small (about 5 MB)
- Support setting the address for the REST API socket
- Drive-by: Create the volume used for postgres from docker-compose

## 📜 Does this require a change to the docs?

- Instructions were added to the README on how to build/run/stop the docker container

## 🐾 Next steps

- Add the container to `packages/hash/external_services`
- Host the docker image so it is not needed to rebuild it outside of the CI

## 🛡 What tests cover this?

- The REST API is now tested using the Docker container

## ❓ How to test this?

- Go to `packages/graph/hash_graph`
- Run the commands provided in the README

### CODE REVIEW

File: .github/scripts/rust/setup.py
Comment (TimDiekmann) [inline .github/scripts/rust/setup.py:17]:
```suggestion
# Build a docker container for these crates
```

File: packages/graph/deployment/graph/Dockerfile
Comment (TimDiekmann) [reply packages/graph/deployment/graph/Dockerfile:23]: `FROM scratch` does not have user handling, the container only contains the binary, **nothing** else.

File: packages/graph/deployment/graph/Dockerfile
Comment (TimDiekmann) [reply packages/graph/deployment/graph/Dockerfile:10]: Alpine images do not have `bash`, so this lint does not apply here. In case this step is failing, a later step will fail as well.

File: packages/graph/deployment/graph/Dockerfile
Comment (TimDiekmann) [reply packages/graph/deployment/graph/Dockerfile:12]: `rust` is a layer created before.

File: .github/workflows/rust.yml
Comment (TimDiekmann) [inline .github/workflows/rust.yml:28]: `bench` was unused (it was used a while ago before we switched to `cargo make`)

File: packages/graph/deployment/docker-compose.yml
Comment (thehabbos007) [inline packages/graph/deployment/docker-compose.yml:4]: Does making docker-compose take care of this volume mean it gets destroyed when we "down" the deployment?

File: packages/graph/hash_graph/README.md
Comment (thehabbos007) [inline packages/graph/hash_graph/README.md:3]: the readme one level up seems tobe similar to this but only contain postgres steps. Which Readme should we make use of? both contents seems to be related to the Graph API

File: packages/graph/deployment/docker-compose.yml
Comment (benwerner01) [inline packages/graph/deployment/docker-compose.yml:4]: I don't have issues with this, but I'm curious why this change was made as part of this PR? Was it added accidentally previously, or has the thinking around it changed somehow?

**Review by thehabbos007 — COMMENTED** [review]
A couple comments, but looks good :+1:

File: packages/graph/deployment/docker-compose.yml
Comment (TimDiekmann) [reply packages/graph/deployment/docker-compose.yml:4]: No, it will only create the volume if it does not exist. It's probably called `deployment_hash-graph-data` (based on the directory it's in).

**Review by benwerner01 — COMMENTED** [review]
Thanks Tim, the docker container is building and running for me locally.

It's probably best to wait for @thehabbos007 or @Alfred-Mountfield's approval of the rust setup changes.

File: packages/graph/deployment/docker-compose.yml
Comment (TimDiekmann) [reply packages/graph/deployment/docker-compose.yml:4]: It's a QoL enhancement, so the caller does not need to create the volume themselves, it will be created (not destroyed) if it's not already existing.

File: packages/graph/hash_graph/README.md
Comment (TimDiekmann) [reply packages/graph/hash_graph/README.md:3]: Yes, the READMEs are confusing, it's not really foreseeable what is written where. We should probably simply merge these REEADMEs in a follow-up PR.

The README one directory up is only for setting up the database, this README describes how to run the Graph on the existing database.

Comment (thehabbos007) [reply packages/graph/hash_graph/README.md:3]: Yes, let's consider this in a follow-up. I suppose it's not really urgent, though

File: packages/graph/hash_graph/README.md
Comment (thehabbos007) [inline packages/graph/hash_graph/README.md:22]: While this does stop the container, the container will still be around. Do we want an instruction to stop and remove the container also? that would be something like `docker rm -f graph` I believe.

File: packages/graph/hash_graph/Makefile.toml
Comment (thehabbos007) [inline packages/graph/hash_graph/Makefile.toml:76]: Since the readme shows using stop instead of kill now, do we want to change this to be stop too?

File: packages/graph/hash_graph/README.md
Comment (TimDiekmann) [reply packages/graph/hash_graph/README.md:22]: Good point, added that to the readme.

File: packages/graph/hash_graph/Makefile.toml
Comment (TimDiekmann) [reply packages/graph/hash_graph/Makefile.toml]: Applied in 02ff56a59b2b90ba1878e0f6e4741f8b550fff55

File: packages/graph/hash_graph/README.md
Comment (thehabbos007) [inline packages/graph/hash_graph/README.md:8]: Was greeted with
`the --mount option requires BuildKit. Refer to https://docs.docker.com/go/buildkit/ to learn how to build images with BuildKit enabled`
when trying to build locally. We should probably add the ENV var in front of the build and maybe refer to https://docs.docker.com/develop/develop-images/build_enhancements/#to-enable-buildkit-builds for enabling buildkit by default

File: packages/graph/hash_graph/Makefile.toml
Comment (thehabbos007) [inline packages/graph/hash_graph/Makefile.toml:59]: Same as above, buildkit isn't enabled by default for me :)

**Review by thehabbos007 — DISMISSED** [review]
Works just fine for me both with cargo make and with the raw docker commands. Thanks :+1:

**Review by thehabbos007 — APPROVED** [review]

**Pull request merged**
